### ISSUE TITLE
Unable to read REST configuration from file: src\main\ml-modules\rest-properties.json 

### ISSUE THREAD

**Issue by sammefford** [sastafford/marklogic-spring-batch#53]
I get the following error when I call 'gradle mlDeploy' with the following gradle.properties:

```
# Publishing configuration
name=data-migrator
group=com.marklogic
version=0.1

# Configuration for the test app
mlHost=localhost
mlAppName=spring-batch
mlRestPort=8200
mlTestRestPort=8201
mlUsername=admin
mlPassword=admin
mlRestAdminUsername=admin
mlRestAdminPassword=admin
mlManageUsername=admin
mlManagePassword=admin
mlAdminUsername=admin
mlAdminPassword=admin
mlContentForestsPerHost=1
```

```
14:25:33.447 [DEBUG] [com.marklogic.client.modulesloader.impl.PropertiesModuleManager] Loading properties from: E:\installs\marklogic\marklogic-spring-batch-0.2.4-alpha\build\ml-last-configured-timestamps.properties
14:25:33.454 [INFO] [com.marklogic.client.modulesloader.impl.DefaultModulesLoader] Writing REST server configuration
14:25:33.454 [INFO] [com.marklogic.client.modulesloader.impl.DefaultModulesLoader] Default document read transform: null
14:25:33.454 [INFO] [com.marklogic.client.modulesloader.impl.DefaultModulesLoader] Transform all documents on read: false
14:25:33.454 [INFO] [com.marklogic.client.modulesloader.impl.DefaultModulesLoader] Validate query options: true
14:25:33.454 [INFO] [com.marklogic.client.modulesloader.impl.DefaultModulesLoader] Validate queries: true
14:25:33.455 [INFO] [com.marklogic.client.modulesloader.impl.DefaultModulesLoader] Output debugging: true
14:25:33.455 [INFO] [com.marklogic.client.modulesloader.impl.DefaultModulesLoader] Update policy: VERSION_OPTIONAL
14:25:33.455 [INFO] [com.marklogic.client.impl.ServerConfigurationManagerImpl] Writing server configuration
14:25:33.456 [DEBUG] [com.marklogic.client.impl.JerseyServices] Putting config/properties/null
14:25:33.513 [DEBUG] [org.apache.http.impl.conn.tsccm.ThreadSafeClientConnManager] Get connection: {}->http://localhost:8200, timeout = 0
14:25:33.513 [DEBUG] [org.apache.http.impl.conn.tsccm.ConnPoolByRoute] [{}->http://localhost:8200] total kept alive: 0, total issued: 0, total allocated: 0 out of 200
14:25:33.513 [DEBUG] [org.apache.http.impl.conn.tsccm.ConnPoolByRoute] No free connections [{}->http://localhost:8200][null]
14:25:33.513 [DEBUG] [org.apache.http.impl.conn.tsccm.ConnPoolByRoute] Available capacity: 100 out of 100 [{}->http://localhost:8200][null]
14:25:33.513 [DEBUG] [org.apache.http.impl.conn.tsccm.ConnPoolByRoute] Creating new connection [{}->http://localhost:8200]
14:25:33.544 [DEBUG] [org.apache.http.impl.conn.DefaultClientConnectionOperator] Connecting to localhost:8200
14:25:34.560 [DEBUG] [org.apache.http.impl.conn.DefaultClientConnectionOperator] Connect to localhost:8200 timed out. Connection will be retried using another IP address
14:25:34.560 [DEBUG] [org.apache.http.impl.conn.DefaultClientConnectionOperator] Connecting to localhost:8200
14:25:35.575 [DEBUG] [org.apache.http.impl.conn.DefaultClientConnection] Connection org.apache.http.impl.conn.DefaultClientConnection@53d9af1 closed
14:25:35.575 [DEBUG] [org.apache.http.impl.conn.DefaultClientConnection] Connection org.apache.http.impl.conn.DefaultClientConnection@53d9af1 shut down
14:25:35.575 [DEBUG] [org.apache.http.impl.conn.tsccm.ThreadSafeClientConnManager] Released connection is not reusable.
14:25:35.575 [DEBUG] [org.apache.http.impl.conn.tsccm.ConnPoolByRoute] Releasing connection [{}->http://localhost:8200][null]
14:25:35.575 [DEBUG] [org.apache.http.impl.conn.DefaultClientConnection] Connection org.apache.http.impl.conn.DefaultClientConnection@53d9af1 closed
14:25:35.575 [DEBUG] [org.apache.http.impl.conn.tsccm.ConnPoolByRoute] Notifying no-one, there are no waiting threads
14:25:35.575 [INFO] [com.marklogic.client.impl.DatabaseClientImpl] Releasing connection
14:25:35.575 [DEBUG] [com.marklogic.client.impl.JerseyServices] Releasing connection
14:25:35.575 [DEBUG] [org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter] Finished executing task ':mlDeployApp'
14:25:35.575 [LIFECYCLE] [class org.gradle.TaskExecutionLogger] :mlDeployApp FAILED
14:25:35.575 [INFO] [org.gradle.execution.taskgraph.AbstractTaskPlanExecutor] :mlDeployApp (Thread[main,5,main]) completed. Took 20.143 secs.
14:25:35.575 [DEBUG] [org.gradle.execution.taskgraph.AbstractTaskPlanExecutor] Task worker [Thread[main,5,main]] finished, busy: 20.205 secs, idle: 0.0 secs
14:25:35.591 [ERROR] [org.gradle.BuildExceptionReporter]
14:25:35.591 [ERROR] [org.gradle.BuildExceptionReporter] FAILURE: Build failed with an exception.
14:25:35.591 [ERROR] [org.gradle.BuildExceptionReporter]
14:25:35.591 [ERROR] [org.gradle.BuildExceptionReporter] * What went wrong:
14:25:35.591 [ERROR] [org.gradle.BuildExceptionReporter] Execution failed for task ':mlDeployApp'.
14:25:35.591 [ERROR] [org.gradle.BuildExceptionReporter] > Unable to read REST configuration from file: E:\installs\marklogic\marklogic-spring-batch-0.2.4-alpha\src\main\ml-modules\rest-properties.json
14:25:35.591 [ERROR] [org.gradle.BuildExceptionReporter]
14:25:35.591 [ERROR] [org.gradle.BuildExceptionReporter] * Try:
14:25:35.591 [ERROR] [org.gradle.BuildExceptionReporter] Run with --stacktrace option to get the stack trace.
14:25:35.591 [LIFECYCLE] [org.gradle.BuildResultLogger]
14:25:35.591 [LIFECYCLE] [org.gradle.BuildResultLogger] BUILD FAILED
14:25:35.591 [LIFECYCLE] [org.gradle.BuildResultLogger]
14:25:35.591 [LIFECYCLE] [org.gradle.BuildResultLogger] Total time: 28.831 secs
14:25:35.622 [DEBUG] [org.gradle.cache.internal.DefaultCacheAccess] Cache task history cache (E:\installs\marklogic\marklogic-spring-batch-0.2.4-alpha\.gradle\2.12\taskArtifacts) was closed 0 times.
14:25:35.622 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Releasing lock on cp_proj class cache for build file 'E:\installs\marklogic\marklogic-spring-batch-0.2.4-alpha\build.gradle' (C:\Users\Sam\.gradle\caches\2.12\scripts\build_b0a1q6p13n5dqo2n98dhbvce3\cp_proj).
14:25:35.622 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Releasing lock on proj class cache for build file 'E:\installs\marklogic\marklogic-spring-batch-0.2.4-alpha\build.gradle' (C:\Users\Sam\.gradle\caches\2.12\scripts\build_b0a1q6p13n5dqo2n98dhbvce3\proj).
14:25:35.622 [DEBUG] [org.gradle.api.internal.artifacts.ivyservice.ivyresolve.memcache.InMemoryCachedRepositoryFactory] In-memory dependency metadata cache closed. Repos cached: 1, cache instances: 1, modules served from cache: 0, artifacts: 0
14:25:35.622 [DEBUG] [org.gradle.api.internal.artifacts.ivyservice.resolveengine.store.CachedStoreFactory] Resolution result cache closed. Cache reads: 0, disk reads: 0 (avg: 0.0 secs, total: 0.0 secs)
14:25:35.622 [DEBUG] [org.gradle.api.internal.artifacts.ivyservice.resolveengine.store.CachedStoreFactory] Resolution result cache closed. Cache reads: 0, disk reads: 0 (avg: 0.0 secs, total: 0.0 secs)
14:25:35.622 [DEBUG] [org.gradle.api.internal.artifacts.ivyservice.resolveengine.store.ResolutionResultsStoreFactory] Deleted 2 resolution results binary files in 0.0 secs
14:25:35.622 [DEBUG] [org.gradle.cache.internal.btree.BTreePersistentIndexedCache] Closing cache module-metadata.bin (C:\Users\Sam\.gradle\caches\modules-2\metadata-2.16\module-metadata.bin)
14:25:35.622 [DEBUG] [org.gradle.cache.internal.btree.BTreePersistentIndexedCache] Closing cache artifact-at-repository.bin (C:\Users\Sam\.gradle\caches\modules-2\metadata-2.16\artifact-at-repository.bin)
14:25:35.622 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Releasing lock on artifact cache (C:\Users\Sam\.gradle\caches\modules-2).
14:25:35.638 [DEBUG] [org.gradle.cache.internal.btree.BTreePersistentIndexedCache] Closing cache plugin-use-metadata.bin (C:\Users\Sam\.gradle\caches\2.12\plugin-resolution\plugin-use-metadata.bin)
14:25:35.638 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Releasing lock on Plugin Resolution Cache (C:\Users\Sam\.gradle\caches\2.12\plugin-resolution).
14:25:35.638 [DEBUG] [org.gradle.api.internal.tasks.compile.daemon.CompilerDaemonManager] Stopping 0 compiler daemon(s).
14:25:35.638 [INFO] [org.gradle.api.internal.tasks.compile.daemon.CompilerDaemonManager] Stopped 0 compiler daemon(s).
```

I can see that I have the file in question:

```
$ ls src/main/ml-modules/ -al
total 1
drwxrwx---+ 1 Sam None   0 May 18 14:26 .
drwxrwx---+ 1 Sam None   0 May 18 13:33 ..
drwxrwx---+ 1 Sam None   0 May 18 13:33 ext
-rwxrwx---+ 1 Sam None 157 May 18 13:33 rest-properties.json
```

**Comment by sastafford** [comment]
I've seen this occasionally with ml-gradle projects.   I've never been able to pinpoint why this happens.  @rjrudin have you seen this before?

**Comment by rjrudin** [comment]
I just ran mlDeploy locally on master, and it worked fine. Ran mlReloadModules a few times, worked fine.

@sammefford  Was there anything in ErrorLog.txt that reported an error?

@sastafford Do you need this file? This is just for setting up a REST API server for testing purposes, right?

**Comment by sastafford** [comment]
I am setting the update policy to VERSION_OPTIONAL because I am utilizing DocumentDescriptors in the Java Client API.  Worst case, I could just set that through the Java Client API.

**Comment by sammefford** [comment]
I wonder if the problem is associated with me running ml-gradle from cygwin

**Comment by sammefford** [comment]
Nevermind, I get the same error when running it from windows 'cmd' window

**Comment by sammefford** [comment]
Nope, @rjrudin , my ErrorLog.txt is empty:

```
-rw-r--r-- 1 daemon daemon 0 May 19 00:00 /var/opt/MarkLogic/Logs/8201_ErrorLog.txt
-rw-r--r-- 1 daemon daemon 0 May 19 00:00 /var/opt/MarkLogic/Logs/8200_ErrorLog.txt
```
and my root ErrorLog.txt doesn't show anything interesting
```
2016-05-19 08:45:24.413 Info: Saving /var/opt/MarkLogic/Forests/Security/0000006c
2016-05-19 08:45:25.797 Info: Saved 1 MB in 1 sec at 1 MB/sec to /var/opt/MarkLogic/Forests/Security/0000006c
2016-05-19 08:45:25.881 Info: Merging 2 MB from /var/opt/MarkLogic/Forests/Security/0000006b and /var/opt/MarkLogic/Forests/Security/0000006c to /var/opt/MarkLogic/Forests/Security/0000006d, timestamp=14636721233703300
2016-05-19 08:45:27.659 Info: Merged 1 MB in 2 sec at 1 MB/sec to /var/opt/MarkLogic/Forests/Security/0000006d
2016-05-19 08:45:30.435 Info: Deleted 1 MB in 1 sec at 21 MB/sec /var/opt/MarkLogic/Forests/Security/0000006b
2016-05-19 08:45:30.485 Info: Deleted 1 MB in 1 sec at 20 MB/sec /var/opt/MarkLogic/Forests/Security/0000006c
2016-05-19 08:48:33.538 Info: Saving /var/opt/MarkLogic/Forests/App-Services/00000012
2016-05-19 08:48:34.035 Info: Saved 15 MB in 1 sec at 30 MB/sec to /var/opt/MarkLogic/Forests/App-Services/00000012
2016-05-19 08:48:34.083 Info: Merging 16 MB from /var/opt/MarkLogic/Forests/App-Services/00000012 and /var/opt/MarkLogic/Forests/App-Services/00000011 to /var/opt/MarkLogic/Forests/App-Services/00000013, timestamp=14636722805609100
2016-05-19 08:48:34.887 Info: Merged 1 MB in 1 sec at 1 MB/sec to /var/opt/MarkLogic/Forests/App-Services/00000013
2016-05-19 08:48:37.454 Info: Deleted 15 MB in 1 sec at 562 MB/sec /var/opt/MarkLogic/Forests/App-Services/00000012
2016-05-19 08:48:37.471 Info: Deleted 1 MB in 1 sec at 59 MB/sec /var/opt/MarkLogic/Forests/App-Services/00000011
2016-05-19 08:48:39.543 Info: Saving /var/opt/MarkLogic/Forests/spring-batch-modules-1/00000006
2016-05-19 08:48:40.067 Info: Saved 15 MB in 1 sec at 29 MB/sec to /var/opt/MarkLogic/Forests/spring-batch-modules-1/00000006
2016-05-19 08:48:40.107 Info: Merging 16 MB from /var/opt/MarkLogic/Forests/spring-batch-modules-1/00000006 and /var/opt/MarkLogic/Forests/spring-batch-modules-1/00000005 to /var/opt/MarkLogic/Forests/spring-batch-modules-1/00000007, timestamp=14636722805609100
2016-05-19 08:48:40.836 Info: Merged 1 MB in 1 sec at 1 MB/sec to /var/opt/MarkLogic/Forests/spring-batch-modules-1/00000007
2016-05-19 08:48:43.445 Info: Deleted 15 MB in 1 sec at 924 MB/sec /var/opt/MarkLogic/Forests/spring-batch-modules-1/00000006
2016-05-19 08:48:43.462 Info: Deleted 1 MB in 1 sec at 59 MB/sec /var/opt/MarkLogic/Forests/spring-batch-modules-1/00000005
(END)

```

**Comment by rjrudin** [comment]
@sammefford  Maybe try removing a couple of the fields from rest-api.json. Just grasping here.

**Comment by sastafford** [comment]
Concur with @rjrudin.  Try moving rest-properties.json to finish the deployment and then add it back in.  I seem to remember that's how I got through initially.

**Comment by sastafford** [comment]
OBE, closed in #63

**Comment by sparnell** [comment]
I've run into this issue before when the REST server has to use BASIC auth (for example, when using external-security). I find that setting mlAppConfig { restAuthentication="BASIC" } makes this error go away. I have no idea why it points to rest-properties.json as the issue, but I've noticed that several unrelated errors return this same error message.
